### ISSUE TITLE
Mocking calls made from __destruct

### ISSUE THREAD

**Issue by ajant** [phpspec/prophecy#266]
We are using __destruct magic method to close connections to resources. If we try to mock resource class, which is then called in destructor to close connection, test fails, reporting calls which should have been made in destructor were expected, but never made. However, when I created simpler example of the same scenario, everything worked perfectly.

In a scenario when test fails, I've tried to debug code. While my debugging session is still running and my cursor is on a break point inside __destruct method, test has already finished and failed, due to the fact that destruct logic has not been called. Obviosly __destruct is called implicitly, after the test was finished, in some cases, but not always. Is there anything that can be done about this?
Thanx

**Comment by stof** [comment]
Well, if you want to make some assertions on the behavior of ``__destruct``, you must write your tests in a way ensuring that this happens before the end of the test (making sure that there is no circular object graphs preventing the refcount to destruct the object is a first step to achieve this btw)

**Comment by ajant** [comment]
This is my scenario (I'm using PHPUnit):
1. define prophecies
2. set expected exception (code being tested should throw one)
3. invoke logic (which throws exception)
4. test fails
5. __desturct invoked

I have no control over instances after I invoke logic. Any ideas?

**Comment by ajant** [comment]
So I've solved the problem, your answer sent me to the right direction.
In stead of using PHPUnit method setExpectedException and losing control of the code, I've surrounded code invocation by try/catch block. Inside catch block I de-referenced tested class, so that constructor is called, then I manually asserted that exception message was as expected and only then finished the test. A bit more code, but everything works fine.

Thanx for the help.
